### PR TITLE
Implement Content-Type middleware

### PR DESCRIPTION
**Pull request by gpietoso**
Added Content-Type middleware. Code ported from HTTPlug's CT plugin, as so, currently supports JSON and XML.

### CODE REVIEW

File: lib/Middleware/ContentTypeMiddleware.php [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+        $this->skipDetection = $options['skip_detection'];
+        $this->sizeLimit = $options['size_limit'];
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function handleRequest(RequestInterface $request, callable $next)
+    {
+       if (!$request->hasHeader('Content-Type')) {
```
Comment (Nyholm) [inline lib/Middleware/ContentTypeMiddleware.php]: If you reverse this if statement you will not use so much indentation.

File: lib/Middleware/ContentTypeMiddleware.php [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+            }
+
+            if ($this->isJson($stream)) {
+                $request = $request->withHeader('Content-Type', 'application/json');
+                return $next($request);
+            }
+
+            if ($this->isXml($stream)) {
+                $request = $request->withHeader('Content-Type', 'application/xml');
+                return $next($request);
```
Comment (Nyholm) [inline lib/Middleware/ContentTypeMiddleware.php]: This could be removed

File: lib/Middleware/ContentTypeMiddleware.php [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+            }
+
+            if ($this->skipDetection && (null === $streamSize || $streamSize >= $this->sizeLimit)) {
+                return $next($request);
+            }
+
+            if ($this->isJson($stream)) {
+                $request = $request->withHeader('Content-Type', 'application/json');
+                return $next($request);
```
Comment (Nyholm) [inline lib/Middleware/ContentTypeMiddleware.php]: This could be removed

File: lib/Middleware/ContentTypeMiddleware.php [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+            }
+
+            if ($this->isXml($stream)) {
```
Comment (Nyholm) [inline lib/Middleware/ContentTypeMiddleware.php]: Use elseif

File: lib/Middleware/ContentTypeMiddleware.php [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+
+            if (!$stream->isSeekable()) {
+                return $next($request);
+            }
+
+            if (0 === $streamSize) {
+                return $next($request);
+            }
+
+            if ($this->skipDetection && (null === $streamSize || $streamSize >= $this->sizeLimit)) {
```
Comment (Nyholm) [inline lib/Middleware/ContentTypeMiddleware.php]: I’m not sure about this line. 

However, shouldn’t this be one of first thing we do? Checking for “skipDetection”?

Comment (gpietoso) [reply lib/Middleware/ContentTypeMiddleware.php]: Seems so. By the way, after checking "skipDetection", couldn't all the other validations be in one if statement using "or"?

```php
if (null === $streamSize || $streamSize >= $this->sizeLimit || 0 === $streamSize 
       || !$stream->isSeekable()) 
{
      return $next($request);
}
```

File: lib/Middleware/ContentTypeMiddleware.php [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+    /**
+     * {@inheritdoc}
+     */
+    public function handleRequest(RequestInterface $request, callable $next)
+    {
+        if ($this->skipDetection) {
+            return $next($request);
+        }
+
+       if ($request->hasHeader('Content-Type')) {
```
[the remark this answers is not in the document]
Comment (gpietoso) [reply lib/Middleware/ContentTypeMiddleware.php]: Ok.

File: lib/Middleware/ContentTypeMiddleware.php [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+        $this->skipDetection = $options['skip_detection'];
+        $this->sizeLimit = $options['size_limit'];
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function handleRequest(RequestInterface $request, callable $next)
+    {
+        if ($this->skipDetection && $request->hasHeader('Content-Type')) {
```
Comment (Nyholm) [inline lib/Middleware/ContentTypeMiddleware.php]: Should be logic “or”, right?

Comment (gpietoso) [reply lib/Middleware/ContentTypeMiddleware.php]: Ops, my distraction. Sorry, was doing this from the train.

File: lib/Middleware/ContentTypeMiddleware.php [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+     * Determine the size stream limit for which the detection as to be skipped (default to 16Mb).
+     *
+     * @var int
+     */
+    protected $sizeLimit;
+
+    /**
+     * @param array $config {
+     *
+     *     @var bool $skip_detection True skip detection if stream size is bigger than $size_limit.
```
Comment (gpietoso) [inline lib/Middleware/ContentTypeMiddleware.php]: Travis output:
--- Original
+++ New
@@ @@
      *
-     *     @var bool $skip_detection True skip detection if stream size is bigger than $size_limit.
+     *     @var bool $skip_detection true skip detection if stream size is bigger than $size_limit
      *     @var int $size_limit size stream limit for which the detection as to be skipped.
      * }
      */

Isn't in this line the period should be in?

Comment (Nyholm) [reply lib/Middleware/ContentTypeMiddleware.php]: Oh. Sorry my bad. Travis suggest to remove it

**Pull request merged**